lib/Conversion/TorchToLinalg/Linear.cpp

```diff
@@ -612,30 +623,86 @@ class ConvertAtenConvolutionOp : public OpConversionPattern<AtenConvolutionOp> {
           AffineMap::get(/*dimCount=*/resultRank, /*symbolCount=*/0,
                          rewriter.getAffineDimExpr(1), context),
           rewriter.getMultiDimIdentityMap(resultRank)};
-      SmallVector<StringRef> iteratorTypes(resultRank, "parallel");
-      biasInitTensor = rewriter
-                           .create<linalg::GenericOp>(
-                               loc, initTensor.getType(), bias, initTensor,
-                               indexingMaps, iteratorTypes,
-                               [](OpBuilder &b, Location loc, ValueRange args) {
-                                 b.create<linalg::YieldOp>(loc, args[0]);
-                               })
-                           .getResult(0);
+      SmallVector<StringRef> iteratorTypes(resultRank,
+                                           getParallelIteratorTypeName());
+      outputTensor = rewriter
+                         .create<linalg::GenericOp>(
+                             loc, initTensor.getType(), bias, initTensor,
+                             indexingMaps, iteratorTypes,
+                             [](OpBuilder &b, Location loc, ValueRange args) {
+                               b.create<linalg::YieldOp>(loc, args[0]);
+                             })
+                         .getResult(0);
     }
 
     auto stridesAttr = rewriter.getI64VectorAttr(strideInts);
     auto dilationAttr = rewriter.getI64VectorAttr(dilationInts);
 
-    // TODO: add 1D and 3D case
-    Value conv =
-        rewriter
-            .create<linalg::Conv2DNchwFchwOp>(
-                loc, biasInitTensor.getType(), ValueRange{paddedInput, weight},
-                biasInitTensor, stridesAttr, dilationAttr)
-            .getResult(0);
+    Value inputStride =
+        rewriter.create<arith::FloorDivSIOp>(loc, inChannels, groups);
+    Value weightStride =
+        rewriter.create<arith::FloorDivSIOp>(loc, weightBatch, groups);
+
+    SmallVector<Value> zeroOffsets(inRank, rewriter.create<arith::ConstantOp>(
+                                               loc, rewriter.getIndexAttr(0)));
+    SmallVector<Value> unitStrides(inRank, rewriter.create<arith::ConstantOp>(
+                                               loc, rewriter.getIndexAttr(1)));
+    SmallVector<Value> outDimSlice(outDims);
+    outDimSlice[1] = weightStride;
+    SmallVector<Value> inputSliceSizes{inBatch, inputStride};
+    inputSliceSizes.append(inDims);
+    SmallVector<Value> weightSliceSizes{weightStride, weightChannels};
+    weightSliceSizes.append(weightDims);
+
+    // Pad the input tensor according to padding.
+    SmallVector<int64_t, 4> paddingIncludingNC = {0, 0};
+    paddingIncludingNC.append(paddingInts);
+
+    for (auto group = 0; group < groupSize; group++) {
```

Comment:
I don't think this is the right approach. This can create a really large amount of IR. We should preserve the fact that this is a grouped convolution rather than splitting it into numGroups individual convolutions. If necessary, we might need to add a linalg named op for this.

Reply:
You might want to look at how depthwise convolution is handled here: https://github.com/tensorflow/mlir-hlo/blob/82c3f9953083be894f88f3253d0434b3cd7490e6/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc#L2076

Reply:
For now, handling just non-grouped and depthwise seems fine. More general groups counts will be more complex and will probably require a linalg named op or something -- we need to see what backends want for that.

Reply:
Ah, I see. Okay, I'll look into the models that depend on this patch and see how pressing the need for non-depthwise grouped convolutions is.

Reply:
> For now, handling just non-grouped and depthwise seems fine. More general groups counts will be more complex and will probably require a linalg named op or something -- we need to see what backends want for that.

Hi @silvasean, can we get this patch merged as of now? Since it will unblock some of the models required by us. @gpetters94 will continue working on your suggestions and modify this patch or add an op in the Linalg dialect.

Reply:
Ouch. Agree it would be good to keep the quality bar and we can carry the patch locally. It would be nice to get this kind of feedback earlier so we don't abandon a patchset after two weeks. Please let us know if there is a better way to front load this kind of feedback. I think @gpetters94 was going off @cathyzhyi reference implementation pointed. I will continue to push for an RFC before an implementation.

Reply:
I can try my hand at implementing a named op in LLVM. Do you think the same logic I'm using here (splitting grouped convs into groups of convs) will be appropriate at the linalg level?

Reply:
Adding a special case for depthwise might also need a new op upstream, since the current ops don't support NchwFchw order. I could do three transpositions but I don't think that's the best way to do it.

Reply:
Ah, yeah, then we should add the new op upstream too to avoid the transpositions.